gcloud/datastore/query.py

```diff
@@ -140,441 +243,266 @@ def filter(self, property_name, operator, value):
-                       a given kind.
-
-        :rtype: :class:`Query` or `int`.
-        :returns: If no arguments, returns the current offset.
-                  If an offset is provided, returns a clone of the
-                  :class:`Query` with that offset set.
+        :rtyoe: tuple, (entities, more_results, cursor)
         """
-        if offset is None:
-            return self._offset
+        pb = _pb_from_query(self._query)
 
-        clone = self._clone()
-        clone._offset = offset
-        clone._pb.offset = offset
-        return clone
+        start_cursor = self._start_cursor
+        if start_cursor is not None:
+            pb.start_cursor = base64.b64decode(start_cursor)
 
-    def group_by(self, group_by=None):
-        """Adds a group_by to the query.
+        end_cursor = self._end_cursor
+        if end_cursor is not None:
+            pb.end_cursor = base64.b64decode(end_cursor)
 
-        This is a hybrid getter / setter, used as::
+        pb.limit = self._limit
+        pb.offset = self._offset
 
-          >>> query = Query('Person')
-          >>> query.group_by()  # Get the group_by for this query.
-          []
-          >>> query = query.group_by(['name'])
-          >>> query.group_by()  # Get the group_by for this query.
-          ['name']
-
-        :type group_by: sequence of strings
-        :param group_by: Each value is a string giving the name of a
-                         property to use to group results together.
+        query_results = self._query.dataset.connection().run_query(
+            query_pb=pb,
+            dataset_id=self._query.dataset.id(),
+            namespace=self._query.namespace,
+            )
+        # NOTE: `query_results` contains an extra value that we don't use,
+        #       namely `skipped_results`.
+        #
+        # NOTE: The value of `more_results` is not currently useful because
+        #       the back-end always returns an enum
+        #       value of MORE_RESULTS_AFTER_LIMIT even if there are no more
+        #       results. See
+        #       https://github.com/GoogleCloudPlatform/gcloud-python/issues/280
+        #       for discussion.
+        entity_pbs, cursor_as_bytes, more_results_enum = query_results[:3]
 
-        :rtype: :class:`Query` or `list` of strings.
-        :returns: If no arguments, returns the current group_by.
-                  If a list of group by properties is provided, returns a clone
-                  of the :class:`Query` with that list of values set.
-        """
-        if group_by is None:
-            return [prop_ref.name for prop_ref in self._pb.group_by]
+        self._start_cursor = base64.b64encode(cursor_as_bytes)
+        self._end_cursor = None
 
-        clone = self._clone()
+        if more_results_enum == self._NOT_FINISHED:
+            self._more_results = True
+        elif more_results_enum in self._FINISHED:
+            self._more_results = False
+        else:   # pragma: NO COVER
```

Comment:
I used to have a test case where we just faked this branch. No need for the "NO COVER".